### PR TITLE
JAMES-3995 Optimize Email/get

### PR DESCRIPTION
**Pull request by chibenwa**
On many very large message we might consume excessive memory.

Avoid buffering of the binary parts in order to save memory.

@Arsnael help wanted to help polish this.

Apparently I did broke some stuff in bodySTructure

### CODE REVIEW

**Comment by Arsnael** [comment]
Alright let me have a look...

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/mime4j/JamesBodyFactory.java [hunk 1]
```diff
@@ -0,0 +1,366 @@
...
+        CountingOutputStream out = new CountingOutputStream(OutputStream.nullOutputStream());
+        is.transferTo(out);
+        return new FakeBinaryBody(out.getCount());
+    }
+
+    public BinaryBody binaryBody(final byte[] buf) {
+        return new BinaryBody1(buf);
+    }
+
+    static class StringBody1 extends TextBody {
```
Comment (Arsnael) [inline server/data/data-jmap/src/main/java/org/apache/james/jmap/mime4j/JamesBodyFactory.java:156]: We should probably find better name for those static classes than `1`, `2`, ...

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/mime4j/SizeUtils.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        }
+        System.out.println(body.getClass());
+        CountingOutputStream countingOutputStream = new CountingOutputStream(OutputStream.nullOutputStream());
+        DefaultMessageWriter defaultMessageWriter = new DefaultMessageWriter();
+        defaultMessageWriter.writeBody(body, countingOutputStream);
+        return countingOutputStream.getCount();
+    }
+
+    /*
+      private def size(entity: Entity): Try[Size] = {
```
Comment (Arsnael) [inline server/data/data-jmap/src/main/java/org/apache/james/jmap/mime4j/SizeUtils.java:70]: Commented code block spotted

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailGetMethodContract.scala [hunk 3]
```diff
@@ -4591,7 +4591,7 @@ trait EmailGetMethodContract {
       .header(ACCEPT.toString, ACCEPT_RFC8621_VERSION_HEADER)
       .body(request)
     .when
-      .post
+      .post.prettyPeek()
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailGetMethodContract.scala:4594]: debug spotted

**Comment by Arsnael** [comment]
Fixed the build

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailGetMethodContract.scala (hunk 3 above)
Comment (Arsnael) [reply server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailGetMethodContract.scala]: Note: yes scalafix throws on that... ^^'

**Comment by Arsnael** [comment]
Alright get it email get contract tests failing too...

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/mail/Email.scala [hunk 4]
```diff
@@ -610,7 +610,7 @@ private class EmailFullViewFactory @Inject()(zoneIdProvider: ZoneIdProvider, pre
         .getOrElse(Failure(new IllegalArgumentException("No message supplied")))
       mime4JMessage <- Email.parseAsMime4JMessage(firstMessage)
       blobId <- BlobId.of(messageId)
-      bodyStructure <- EmailBodyPart.ofMessage(request.bodyProperties, zoneIdProvider.get(), blobId, mime4JMessage, firstMessage.getSize)
+      bodyStructure <- EmailBodyPart.ofMessage(request.bodyProperties, zoneIdProvider.get(), blobId, mime4JMessage, firstMessage.getBody.size)
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/mail/Email.scala:613]: Was putting the full message size for the body one

Reduced failing tests from 5 to 1 on EmailGetMethodContract

There is still a problem with `bodyStructureForSimpleMultipart` test, size difference of 2 (actual 2285, expected 2287). Dont know if it's right and previous code wrong or if there is a little trick here.

Will check back after lunch

Comment (ottoka) [reply server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/mail/Email.scala:613]: Just a guess, maybe someone counts the CR+LF between headers and body towards the body size?

Comment (Arsnael) [reply server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/mail/Email.scala:613]: Maybe there is something like that... Honestly could not put my finger on it.
Thanks for having a guess at it though :)

I think I will pretend it's acceptable. @chibenwa don't hesitate to disagree with this decision if you think it's better

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/mime4j/SizeUtils.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+import org.apache.james.mime4j.dom.Entity;
+import org.apache.james.mime4j.dom.SingleBody;
+import org.apache.james.mime4j.message.BodyPart;
+import org.apache.james.mime4j.message.DefaultMessageWriter;
+import org.apache.james.mime4j.message.MessageImpl;
+
+import com.google.common.io.CountingOutputStream;
+
+public class SizeUtils {
+    public static long sizeOf(Entity entity) throws IOException {
```
Comment (chibenwa) [inline server/data/data-jmap/src/main/java/org/apache/james/jmap/mime4j/SizeUtils.java:35]: The `Email/get` test suite was imcomplete. I got on a production workload a call for `MultipartImp` with some FakeBinaryBody in it.

Regarding the JMAP spec, `partId` and `blobId` are null for multiparts so there is no downloads.

To quote the spec:

```
(as referenced by the blobId, i.e., the number of octets in the file the user would download).
```

So I fixed this by returning `0` for multiparts as they cannot be downloaded...

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/mime4j/SizeUtils.java (hunk 2 above)
Comment (chibenwa) [reply server/data/data-jmap/src/main/java/org/apache/james/jmap/mime4j/SizeUtils.java:70]: Can be removed.

Sorry I was in a hurry

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailGetMethodContract.scala [hunk 6]
```diff
@@ -3865,7 +3865,7 @@ trait EmailGetMethodContract {
          |                                    "value": "$contentType"
          |                                }
          |                            ],
-         |                            "size": 2287,
+         |                            "size": 2285,
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailGetMethodContract.scala:3868]: As said above the only one reason this do not blow up is that the parts are text body and not binary body...

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/EmailGetMethod.scala [hunk 7]
```diff
@@ -89,6 +89,7 @@ class EmailGetMethod @Inject() (readerFactory: EmailViewReaderFactory,
       case e: IllegalArgumentException => SMono.just(Invocation.error(ErrorCode.InvalidArguments, e.getMessage, invocation.invocation.methodCallId))
       case e: Throwable => SMono.error(e)
     }).map(invocationResult => InvocationWithContext(invocationResult, invocation.processingContext))
+      .doOnError(e => e.printStackTrace())
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/EmailGetMethod.scala:92]: Oups

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/mail/Email.scala (hunk 4 above)
Comment (chibenwa) [reply server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/mail/Email.scala:613]: Quite likely.

However that's the size of the message as stored in the mailbox.

I suggest downloading the EML and checking the size and validating which on it the right one...

It is not impossible the the MessageWriter does some sanitizing like adding a trailing CRLF if missing...

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/mime4j/JamesBodyFactory.java [hunk 8]
```diff
@@ -0,0 +1,366 @@
...
+import org.apache.james.mime4j.message.BodyFactory;
+import org.apache.james.mime4j.util.ByteArrayOutputStreamRecycler;
+import org.apache.james.mime4j.util.ContentUtil;
+
+import com.google.common.io.CountingOutputStream;
+
+/**
+ * Factory for creating message bodies.
+ */
+public class JamesBodyFactory implements BodyFactory {
```
Comment (chibenwa) [inline server/data/data-jmap/src/main/java/org/apache/james/jmap/mime4j/JamesBodyFactory.java:48]: Better name needed: how about `AvoidBinaryBodyBufferingBodyFactory` ?

**Comment by Arsnael** [comment]
Seems creating issues with Email/parse method as well: https://ci-builds.apache.org/job/james/job/ApacheJames/job/PR-2005/7/testReport/

Will check...

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/mail/Email.scala (hunk 4 above)
Comment (Arsnael) [reply server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/mail/Email.scala:613]: Might be that... seems to be 2285 to me when I download the eml and check the multipart/mixed part. Maybe some sanitizing at the work yes. So I'm fine with this honestly

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/EmailParseMethod.scala [hunk 9]
```diff
@@ -103,7 +103,7 @@ class EmailParseMethod @Inject()(val blobResolvers: BlobResolvers,
   private def parse(request: EmailParseRequest, blobId: BlobId, blobContent: => InputStream, size: Size): SMono[EmailParseResults] = {
     val result: SMono[EmailParseView] = for {
       mime4JMessage <- SMono.fromTry(Email.parseStreamAsMime4JMessage(blobContent))
-      bodyStructure <- SMono.fromTry(EmailBodyPart.of(request.bodyProperties, zoneIdProvider.get(), blobId, mime4JMessage))
+      bodyStructure <- SMono.fromTry(EmailBodyPart.ofMessage(request.bodyProperties, zoneIdProvider.get(), blobId, mime4JMessage, size))
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/EmailParseMethod.scala:106]: <3 cool catch thanks Rene!

Comment (Arsnael) [reply server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/EmailParseMethod.scala:106]: Well the Email/parse test were failing without this :)

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/mime4j/SizeUtils.java (hunk 5 above)
Comment (chibenwa) [reply server/data/data-jmap/src/main/java/org/apache/james/jmap/mime4j/SizeUtils.java:35]: Is this handled?

Comment (Arsnael) [reply server/data/data-jmap/src/main/java/org/apache/james/jmap/mime4j/SizeUtils.java:35]: When I read your comment, I thought you did, as it finished with a sentence `So I fixed this...`

Did I misunderstand something?

**Comment by Arsnael** [comment]
> Could we tripple check exotic multiparts?

I ain't too sure about all of that honestly, but could try

**Comment by chibenwa** [comment]
Email/get on these two mails:

[multipart_report.txt](https://github.com/apache/james-project/files/14350673/multipart_report.txt)

[mixed-alternative.txt](https://github.com/apache/james-project/files/14350758/mixed-alternative.txt)

Let's see if this pass!

And I insist that if we call size on a multipart, we return explicitly `0`...

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/mime4j/SizeUtils.java (hunk 5 above)
Comment (Arsnael) [reply server/data/data-jmap/src/main/java/org/apache/james/jmap/mime4j/SizeUtils.java:35]: Well then you need to be more clear in your comments then. 

Will take a look, with your suggested tests for other multiparts that seem to test this

**Comment by chibenwa** [comment]
I found the faulty input!

[nested_mixed_alternative.txt](https://github.com/apache/james-project/files/14365621/nested_mixed_alternative.txt)

That's a pretty big deal... Because we need to give a size for the nested message that could be downloaded. `0` is actually not the answer.

More work needed here. I suggest making the mime4J DefaultMessageWriter able to compute size...

**Comment by chibenwa** [comment]
Edit: I did find my way around?

**Comment by Arsnael** [comment]
```
02:47:02,846 [ERROR] Failed to execute goal io.github.evis:scalafix-maven-plugin_2.13:0.1.8_0.11.0:scalafix (scala-check-style) on project jmap-rfc-8621-integration-tests-common: Scalafix invoked with errors. Check logs for details. -> [Help 1]
```

**Comment by chibenwa** [comment]
Some fine tunning needed apparently...

`org.apache.james.jmap.rfc8621.memory.MemoryEmailGetMethodTest.shouldSupportAttachedMessageWithComplexMultipart(GuiceJamesServer)`

**Review by Arsnael — APPROVED** [review]

**Pull request merged**